From e614730587b053823453f65cd53cad168dd4952a Mon Sep 17 00:00:00 2001
From: c4pp4
Date: Thu, 22 Oct 2026 04:30:58 +0200
Subject: [PATCH 1/1] Classic non-transparent panel with shadow

Make the top panel non-transparent and draw the panel shadow below or
over the windows as in the classic Unity.

Signed-off-by: c4pp4
---
 panel/PanelController.cpp             |  2 +-
 panel/PanelView.cpp                   |  2 +-
 plugins/unityshell/src/unityshell.cpp | 64 +++++++++++++++++++++++++++
 3 files changed, 66 insertions(+), 2 deletions(-)

diff --git a/panel/PanelController.cpp b/panel/PanelController.cpp
index c6496b7..f24eb46 100644
--- a/panel/PanelController.cpp
+++ b/panel/PanelController.cpp
@@ -66,7 +66,7 @@ Controller::Impl::Impl(Controller* parent, menu::Manager::Ptr const& indicators,
   : parent_(parent)
   , indicators_(indicators)
   , edge_barriers_(edge_barriers)
-  , opacity_(0.7f)
+  , opacity_(1.0f)
   , opacity_maximized_toggle_(false)
 {
   UScreen* screen = UScreen::GetDefault();
diff --git a/panel/PanelView.cpp b/panel/PanelView.cpp
index 1a54ab1..89fc9ed 100644
--- a/panel/PanelView.cpp
+++ b/panel/PanelView.cpp
@@ -58,7 +58,7 @@ PanelView::PanelView(MockableBaseWindow* parent, menu::Manager::Ptr const& menus
   , opacity_maximized_toggle_(false)
   , needs_geo_sync_(false)
   , overlay_is_open_(false)
-  , opacity_(0.7f)
+  , opacity_(1.0f)
   , monitor_(0)
   , stored_dash_width_(0)
   , bg_effect_helper_(this)
diff --git a/plugins/unityshell/src/unityshell.cpp b/plugins/unityshell/src/unityshell.cpp
index 38f433a..ff739f6 100644
--- a/plugins/unityshell/src/unityshell.cpp
+++ b/plugins/unityshell/src/unityshell.cpp
@@ -3157,9 +3157,73 @@ bool UnityWindow::glDraw(const GLMatrix& matrix,
 
   auto draw_panel_shadow = DrawPanelShadow::NO;
 
+  if (!(mask & PAINT_WINDOW_ON_TRANSFORMED_SCREEN_MASK))
+  {
+    Window active_window = screen->activeWindow();
+
+    if (G_UNLIKELY(window_type == CompWindowTypeDesktopMask))
+    {
+      uScreen->setPanelShadowMatrix(matrix);
+
+      if (active_window == 0 || active_window == window->id())
+      {
+        if (PluginAdapter::Default().IsWindowOnTop(window->id()))
+        {
+          draw_panel_shadow = DrawPanelShadow::OVER_WINDOW;
+        }
+        uScreen->is_desktop_active_ = true;
+      }
+    }
+    else
+    {
+      if (window->id() == active_window)
+      {
+        draw_panel_shadow = DrawPanelShadow::BELOW_WINDOW;
+        uScreen->is_desktop_active_ = false;
+
+        if (!(window_state & CompWindowStateMaximizedVertMask) &&
+            !(window_state & CompWindowStateFullscreenMask) &&
+            !(window_type & CompWindowTypeFullscreenMask))
+        {
+          auto const& output = uScreen->screen->currentOutputDev();
+          int monitor = uScreen->WM.MonitorGeometryIn(NuxGeometryFromCompRect(output));
+
+          if (window->y() - window->border().top < output.y() + uScreen->panel_style_.PanelHeight(monitor))
+          {
+            draw_panel_shadow = DrawPanelShadow::OVER_WINDOW;
+          }
+        }
+      }
+      else if (uScreen->menus_->integrated_menus())
+      {
+        draw_panel_shadow = DrawPanelShadow::BELOW_WINDOW;
+      }
+      else
+      {
+        if (uScreen->is_desktop_active_)
+        {
+          if (PluginAdapter::Default().IsWindowOnTop(window->id()))
+          {
+            draw_panel_shadow = DrawPanelShadow::OVER_WINDOW;
+            uScreen->panelShadowPainted = CompRegion();
+          }
+        }
+      }
+    }
+  }
+
+  if (locked)
+    draw_panel_shadow = DrawPanelShadow::NO;
+
+  if (draw_panel_shadow == DrawPanelShadow::BELOW_WINDOW)
+    uScreen->paintPanelShadow(region);
+
   deco_win_->Draw(matrix, attrib, region, mask);
   bool ret = gWindow->glDraw(matrix, attrib, region, mask);
 
+  if (draw_panel_shadow == DrawPanelShadow::OVER_WINDOW)
+    uScreen->paintPanelShadow(region);
+
   return ret;
 }
 
-- 
2.55.0

